File: myrm-agent-server/app/database/models/kanban.py
"""
[INPUT] models.base::Base (POS: ORM model base class)
[OUTPUT] KanbanBoardModel: Kanban board, KanbanTaskModel: Kanban task,
        KanbanTaskRunModel: Execution run, KanbanTaskEventModel: Lifecycle event,
        KanbanTaskEdgeModel: Task dependency edge
[POS] Kanban domain ORM models. Manages boards, tasks, runs, events,
and dependency edges for persistent multi-task scheduling.
"""

from datetime import datetime

from sqlalchemy import (
    JSON,
    Boolean,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from .base import Base


class KanbanBoardModel(Base):
    """Kanban board — top-level grouping entity."""

    __tablename__ = "kanban_boards"

    id: Mapped[str] = mapped_column(String(32), primary_key=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str] = mapped_column(Text, default="", nullable=False)

    max_concurrent_tasks: Mapped[int] = mapped_column(
        Integer, default=3, nullable=False
    )
    heartbeat_interval_seconds: Mapped[int] = mapped_column(
        Integer, default=30, nullable=False
    )
    zombie_timeout_seconds: Mapped[int] = mapped_column(
        Integer, default=120, nullable=False
    )
    max_retries_per_task: Mapped[int] = mapped_column(
        Integer, default=3, nullable=False
    )
    auto_block_after_consecutive_failures: Mapped[int] = mapped_column(
        Integer, default=5, nullable=False
    )
    specify_max_tokens: Mapped[int] = mapped_column(
        Integer, default=6000, nullable=False, server_default="6000",
    )
    auto_specify_on_create: Mapped[bool] = mapped_column(
        Boolean, default=False, nullable=False, server_default="0",
    )
    default_workdir: Mapped[str | None] = mapped_column(
        String(1024), nullable=True, default=None,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    tasks: Mapped[list["KanbanTaskModel"]] = relationship(
        "KanbanTaskModel", back_populates="board", cascade="all, delete-orphan"
    )


class KanbanTaskModel(Base):
    """Kanban task — unit of work on a board."""

    __tablename__ = "kanban_tasks"
    __table_args__ = (
        Index("ix_kanban_tasks_board_status", "board_id", "status"),
    )

    id: Mapped[str] = mapped_column(String(32), primary_key=True)
    board_id: Mapped[str] = mapped_column(
        String(32),
        ForeignKey("kanban_boards.id", ondelete="CASCADE"),
        nullable=False,
    )
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    description: Mapped[str] = mapped_column(Text, default="", nullable=False)
    status: Mapped[str] = mapped_column(
        String(20), default="backlog", nullable=False,
    )
    priority: Mapped[str] = mapped_column(
        String(20), default="normal", nullable=False,
    )

    agent_id: Mapped[str | None] = mapped_column(String(255), nullable=True, index=True)
    goal_id: Mapped[str | None] = mapped_column(String(255), nullable=True)
    parent_task_id: Mapped[str | None] = mapped_column(
        String(32), ForeignKey("kanban_tasks.id", ondelete="SET NULL"), nullable=True
    )

    workspace_path: Mapped[str | None] = mapped_column(
        String(1024), nullable=True, default=None,
    )
    branch: Mapped[str | None] = mapped_column(
        String(255), nullable=True, default=None,
    )

    max_runtime_seconds: Mapped[int | None] = mapped_column(
        Integer, nullable=True,
    )

    retry_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    max_retries: Mapped[int] = mapped_column(Integer, default=3, nullable=False)
    consecutive_failures: Mapped[int] = mapped_column(
        Integer, default=0, nullable=False
    )
    block_cycle_count: Mapped[int] = mapped_column(
        Integer, default=0, nullable=False, server_default="0",
    )
    last_heartbeat_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    progress_note: Mapped[str | None] = mapped_column(Text, nullable=True)
    blocked_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    block_kind: Mapped[str | None] = mapped_column(String(20), nullable=True)
    scheduled_until: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True,
    )
    result: Mapped[str] = mapped_column(Text, default="", nullable=False)
    error: Mapped[str] = mapped_column(Text, default="", nullable=False)

    metadata_json: Mapped[dict | None] = mapped_column("metadata", JSON, nullable=True)
    extra_skill_ids_json: Mapped[list | None] = mapped_column(
        "extra_skill_ids", JSON, nullable=True, default=None,
    )
    attachment_ids_json: Mapped[list | None] = mapped_column(
        "attachment_ids", JSON, nullable=True, default=None,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )
    completed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    board: Mapped["KanbanBoardModel"] = relationship(
        "KanbanBoardModel", back_populates="tasks"
    )
    parent: Mapped["KanbanTaskModel | None"] = relationship(
        "KanbanTaskModel",
        back_populates="children",
        remote_side="KanbanTaskModel.id",
        foreign_keys="KanbanTaskModel.parent_task_id",
    )
    children: Mapped[list["KanbanTaskModel"]] = relationship(
        "KanbanTaskModel",
        back_populates="parent",
        foreign_keys="KanbanTaskModel.parent_task_id",
    )
    runs: Mapped[list["KanbanTaskRunModel"]] = relationship(
        "KanbanTaskRunModel", back_populates="task", cascade="all, delete-orphan"
    )
    events: Mapped[list["KanbanTaskEventModel"]] = relationship(
        "KanbanTaskEventModel", back_populates="task", cascade="all, delete-orphan"
    )


class KanbanTaskRunModel(Base):
    """Execution run — independent record per attempt."""

    __tablename__ = "kanban_task_runs"

    id: Mapped[str] = mapped_column(String(32), primary_key=True)
    task_id: Mapped[str] = mapped_column(
        String(32),
        ForeignKey("kanban_tasks.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    worker_id: Mapped[str] = mapped_column(String(255), nullable=False)
    outcome: Mapped[str | None] = mapped_column(String(20), nullable=True)
    summary: Mapped[str] = mapped_column(Text, default="", nullable=False)
    error: Mapped[str] = mapped_column(Text, default="", nullable=False)
    metadata_json: Mapped[dict | None] = mapped_column("metadata", JSON, nullable=True)

    started_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    ended_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    task: Mapped["KanbanTaskModel"] = relationship(
        "KanbanTaskModel", back_populates="runs"
    )


class KanbanTaskEventModel(Base):
    """Lifecycle event — persistent audit trail."""

    __tablename__ = "kanban_task_events"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    task_id: Mapped[str] = mapped_column(
        String(32),
        ForeignKey("kanban_tasks.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    kind: Mapped[str] = mapped_column(String(30), nullable=False)
    payload_json: Mapped[dict | None] = mapped_column("payload", JSON, nullable=True)
    run_id: Mapped[str | None] = mapped_column(String(32), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )

    task: Mapped["KanbanTaskModel"] = relationship(
        "KanbanTaskModel", back_populates="events"
    )


class KanbanTaskEdgeModel(Base):
    """Task dependency edge — parent→child DAG constraint."""

    __tablename__ = "kanban_task_edges"
    __table_args__ = (
        UniqueConstraint("parent_task_id", "child_task_id", name="uq_kanban_edge"),
        Index("ix_kanban_edges_child", "child_task_id"),
        Index("ix_kanban_edges_parent", "parent_task_id"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    parent_task_id: Mapped[str] = mapped_column(
        String(32),
        ForeignKey("kanban_tasks.id", ondelete="CASCADE"),
        nullable=False,
    )
    child_task_id: Mapped[str] = mapped_column(
        String(32),
        ForeignKey("kanban_tasks.id", ondelete="CASCADE"),
        nullable=False,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )

File: myrm-agent-frontend/src/services/kanban.ts
import { apiRequest } from '@/lib/api';

// ==================== Types ====================

export type TaskStatus = 'triage' | 'backlog' | 'ready' | 'running' | 'blocked' | 'completed' | 'failed' | 'archived';
export type TaskPriority = 'low' | 'normal' | 'high' | 'urgent';
export type BlockKind = 'human' | 'scheduled' | 'external';

export interface BoardSettings {
  max_concurrent_tasks: number;
  heartbeat_interval_seconds: number;
  zombie_timeout_seconds: number;
  max_retries_per_task: number;
  auto_block_after_consecutive_failures: number;
  specify_max_tokens: number;
  auto_specify_on_create: boolean;
  default_workdir?: string | null;
}

export interface KanbanBoard {
  board_id: string;
  name: string;
  description: string;
  settings: BoardSettings;
  created_at: string;
  updated_at: string;
}

export interface DiagnosticSummary {
  count: number;
  max_severity: 'warning' | 'error' | 'critical' | null;
}

export interface DiagnosticAction {
  kind: string;
  label: string;
  payload: Record<string, string>;
  suggested: boolean;
}

export interface TaskDiagnostic {
  rule_id: string;
  severity: 'warning' | 'error' | 'critical';
  title: string;
  detail: string;
  actions: DiagnosticAction[];
}

export interface AttachmentInfo {
  file_id: string;
  filename: string;
  content_type: string;
  url: string;
}

export interface KanbanTask {
  task_id: string;
  board_id: string;
  title: string;
  description: string;
  status: TaskStatus;
  priority: TaskPriority;
  agent_id?: string | null;
  goal_id?: string | null;
  parent_task_id?: string | null;
  workspace_path?: string | null;
  branch?: string | null;
  retry_count: number;
  max_retries: number;
  consecutive_failures: number;
  blocked_reason?: string | null;
  block_kind?: BlockKind | null;
  scheduled_until?: string | null;
  progress_note?: string | null;
  result: string;
  error: string;
  metadata: Record<string, unknown>;
  extra_skill_ids: string[];
  attachment_ids: string[];
  attachments: AttachmentInfo[];
  max_runtime_seconds?: number | null;
  completion_criteria?: string | null;
  dep_count: number;
  children_total: number;
  children_done: number;
  comment_count: number;
  diagnostics_summary?: DiagnosticSummary | null;
  created_at: string;
  updated_at: string;
  completed_at?: string | null;
}

export interface AgentTaskCounts {
  agent_id: string | null;
  counts: Record<string, number>;
  total: number;
}

export interface BoardSummary {
  board: KanbanBoard;
  task_counts: Record<string, number>;
  total_tasks: number;
  dispatcher_active: boolean;
  by_agent: AgentTaskCounts[];
  oldest_ready_age_seconds: number | null;
}

// ==================== Board API ====================

export async function listBoards(): Promise<{ items: KanbanBoard[]; total: number }> {
  return apiRequest('/kanban/boards');
}

export async function createBoard(data: {
  name: string;
  description?: string;
  max_concurrent_tasks?: number;
  default_workdir?: string;
}): Promise<KanbanBoard> {
  return apiRequest('/kanban/boards', { method: 'POST', body: JSON.stringify(data) });
}

export async function getBoard(boardId: string): Promise<KanbanBoard> {
  return apiRequest(`/kanban/boards/${boardId}`);
}

export async function updateBoard(
  boardId: string,
  data: { name?: string; description?: string; max_concurrent_tasks?: number; default_workdir?: string | null },
): Promise<KanbanBoard> {
  return apiRequest(`/kanban/boards/${boardId}`, { method: 'PATCH', body: JSON.stringify(data) });
}

export async function deleteBoard(boardId: string): Promise<void> {
  return apiRequest(`/kanban/boards/${boardId}`, { method: 'DELETE' });
}

export async function getBoardSummary(boardId: string): Promise<BoardSummary> {
  return apiRequest(`/kanban/boards/${boardId}/summary`);
}

// ==================== Task API ====================

export async function listTasks(
  boardId: string,
  opts?: { status?: TaskStatus; agent_id?: string; limit?: number; offset?: number },
): Promise<{ items: KanbanTask[]; total: number }> {
  const params = new URLSearchParams();
  if (opts?.status) params.set('status_filter', opts.status);
  if (opts?.agent_id) params.set('agent_id', opts.agent_id);
  if (opts?.limit) params.set('limit', String(opts.limit));
  if (opts?.offset) params.set('offset', String(opts.offset));
  const qs = params.toString();
  return apiRequest(`/kanban/boards/${boardId}/tasks${qs ? `?${qs}` : ''}`);
}

export async function createTask(
  boardId: string,
  data: {
    title: string;
    description?: string;
    priority?: TaskPriority;
    agent_id?: string;
    max_retries?: number;
    depends_on?: string[];
    extra_skill_ids?: string[];
    attachment_ids?: string[];
    completion_criteria?: string;
    max_runtime_seconds?: number;
    initial_status?: TaskStatus;
    workspace_path?: string;
    branch?: string;
  },
): Promise<KanbanTask> {
  return apiRequest(`/kanban/boards/${boardId}/tasks`, { method: 'POST', body: JSON.stringify(data) });
}

export async function getTask(taskId: string): Promise<KanbanTask> {
  return apiRequest(`/kanban/tasks/${taskId}`);
}

export async function updateTask(
  taskId: string,
  data: {
    title?: string;
    description?: string;
    priority?: TaskPriority;
    agent_id?: string | null;
    extra_skill_ids?: string[];
    attachment_ids?: string[];
    max_runtime_seconds?: number | null;
    completion_criteria?: string | null;
    result?: string;
    metadata?: Record<string, unknown>;
  },
): Promise<KanbanTask> {
  return apiRequest(`/kanban/tasks/${taskId}`, { method: 'PATCH', body: JSON.stringify(data) });
}

export async function moveTask(
  taskId: string,
  status: TaskStatus,
  opts?: { force?: boolean; block_kind?: BlockKind; blocked_reason?: string; scheduled_until?: string; result?: string; metadata?: Record<string, unknown> },
): Promise<KanbanTask> {
  const { force = false, ...rest } = opts ?? {};
  return apiRequest(`/kanban/tasks/${taskId}/move`, {
    method: 'POST',
    body: JSON.stringify({ status, force, ...rest }),
  });
}

export interface PromoteResult {
  promoted: boolean;
  forced: boolean;
  reason: string | null;
  unmet_parents: { task_id: string; title: string; status: string }[];
}

export async function promoteTask(taskId: string, force = false, reason?: string): Promise<PromoteResult> {
  return apiRequest(`/kanban/tasks/${taskId}/promote`, {
    method: 'POST',
    body: JSON.stringify({ force, reason: reason ?? null }),
  });
}

export interface ReclaimResult {
  reclaimed: boolean;
  task: KanbanTask | null;
}

export async function reclaimTask(taskId: string, reason?: string, newAgentId?: string): Promise<ReclaimResult> {
  return apiRequest(`/kanban/tasks/${taskId}/reclaim`, {
    method: 'POST',
    body: JSON.stringify({
      reason: reason ?? null,
      new_agent_id: newAgentId ?? null,
    }),
  });
}

export async function deleteTask(taskId: string): Promise<void> {
  return apiRequest(`/kanban/tasks/${taskId}`, { method: 'DELETE' });
}

// ==================== Bulk Actions ====================

export type BulkAction = 'move' | 'archive' | 'reassign' | 'reclaim' | 'delete';

export interface BulkActionResult {
  results: { task_id: string; success: boolean; error?: string }[];
  total: number;
  succeeded: number;
  failed: number;
}

export async function bulkAction(
  boardId: string,
  taskIds: string[],
  action: BulkAction,
  params: Record<string, string> = {},
  confirm = false,
): Promise<BulkActionResult> {
  return apiRequest(`/kanban/boards/${boardId}/tasks/bulk-action`, {
    method: 'POST',
    body: JSON.stringify({ task_ids: taskIds, action, params, confirm }),
  });
}

// ==================== Run & Event Types ====================

export interface TaskRun {
  run_id: string;
  task_id: string;
  worker_id: string;
  started_at: string;
  ended_at?: string | null;
  outcome?: string | null;
  summary: string;
  error: string;
  duration_seconds?: number | null;
}

export interface TaskEvent {
  event_id: number;
  task_id: string;
  kind: string;
  payload?: Record<string, unknown> | null;
  run_id?: string | null;
  created_at: string;
}

// ==================== Run & Event API ====================

export async function listRuns(taskId: string): Promise<{ items: TaskRun[]; total: number }> {
  return apiRequest(`/kanban/tasks/${taskId}/runs`);
}

export async function listEvents(
  taskId: string,
  opts?: { sinceId?: number },
): Promise<{ items: TaskEvent[]; total: number }> {
  const params = new URLSearchParams();
  if (opts?.sinceId != null) params.set('since_id', String(opts.sinceId));
  const qs = params.toString();
  return apiRequest(`/kanban/tasks/${taskId}/events${qs ? `?${qs}` : ''}`);
}

// ==================== Board Events API ====================

export interface BoardEvent {
  event_id: number;
  task_id: string;
  task_title: string;
  task_assignee: string;
  kind: string;
  payload?: Record<string, unknown> | null;
  run_id?: string | null;
  created_at: string;
}

export async function listBoardEvents(
  boardId: string,
  opts?: {
    kinds?: string[];
    assignee?: string;
    sinceId?: number;
    sinceTime?: string;
    limit?: number;
  },
): Promise<{ items: BoardEvent[]; total: number }> {
  const params = new URLSearchParams();
  if (opts?.kinds?.length) params.set('kinds', opts.kinds.join(','));
  if (opts?.assignee) params.set('assignee', opts.assignee);
  if (opts?.sinceId != null) params.set('since_id', String(opts.sinceId));
  if (opts?.sinceTime) params.set('since_time', opts.sinceTime);
  if (opts?.limit != null) params.set('limit', String(opts.limit));
  const qs = params.toString();
  return apiRequest(`/kanban/boards/${boardId}/events${qs ? `?${qs}` : ''}`);
}

// ==================== Comment API ====================

export async function addComment(taskId: string, body: string, author: string = 'user'): Promise<TaskEvent> {
  return apiRequest(`/kanban/tasks/${taskId}/comments`, {
    method: 'POST',
    body: JSON.stringify({ body, author }),
  });
}

// ==================== Dependency API ====================

export interface TaskDependency {
  parent_task_id: string;
  child_task_id: string;
}

export async function listBoardEdges(boardId: string): Promise<{ items: TaskDependency[]; total: number }> {
  return apiRequest(`/kanban/boards/${boardId}/edges`);
}

export async function listDependencies(taskId: string): Promise<{ items: string[]; total: number }> {
  return apiRequest(`/kanban/tasks/${taskId}/dependencies`);
}

export async function listDependents(taskId: string): Promise<{ items: string[]; total: number }> {
  return apiRequest(`/kanban/tasks/${taskId}/dependents`);
}

export async function addDependency(childTaskId: string, parentTaskId: string): Promise<TaskDependency> {
  return apiRequest(`/kanban/tasks/${childTaskId}/dependencies`, {
    method: 'POST',
    body: JSON.stringify({ parent_task_id: parentTaskId }),
  });
}

export async function removeDependency(childTaskId: string, parentTaskId: string): Promise<void> {
  return apiRequest(`/kanban/tasks/${childTaskId}/dependencies/${parentTaskId}`, {
    method: 'DELETE',
  });
}

// ==================== Diagnostics API ====================

export async function getTaskDiagnostics(taskId: string): Promise<{ task_id: string; diagnostics: TaskDiagnostic[] }> {
  return apiRequest(`/kanban/tasks/${taskId}/diagnostics`);
}

// ==================== Specify (TRIAGE → spec rewrite) ====================

export interface SpecifyOutcome {
  task_id: string;
  ok: boolean;
  reason: string;
  new_title: string | null;
  new_body: string | null;
  prompt_tokens: number | null;
  completion_tokens: number | null;
  persisted: boolean;
}

export interface SpecifyAllResult {
  items: SpecifyOutcome[];
  total: number;
  persisted: boolean;
}

export async function specifyTask(taskId: string, opts?: { dryRun?: boolean }): Promise<SpecifyOutcome> {
  const dryRun = opts?.dryRun ?? true;
  return apiRequest(`/kanban/tasks/${taskId}/specify?dry_run=${dryRun}`, {
    method: 'POST',
  });
}

export async function applySpec(
  taskId: string,
  data: {
    new_title: string | null;
    new_body: string;
    prompt_tokens?: number | null;
    completion_tokens?: number | null;
  },
): Promise<SpecifyOutcome> {
  return apiRequest(`/kanban/tasks/${taskId}/apply-spec`, {
    method: 'POST',
    body: JSON.stringify(data),
  });
}

export async function specifyAllTriage(boardId: string, opts?: { dryRun?: boolean }): Promise<SpecifyAllResult> {
  const dryRun = opts?.dryRun ?? true;
  return apiRequest(`/kanban/boards/${boardId}/specify-all?dry_run=${dryRun}`, {
    method: 'POST',
  });
}

// ==================== Decompose (TRIAGE → child task graph) ====================

export interface DecomposeChildSpec {
  title: string;
  body: string;
  assignee: string | null;
  parent_indices: number[];
  extra_skill_ids: string[];
}

export interface DecomposeOutcome {
  task_id: string;
  ok: boolean;
  fanout: boolean;
  children: DecomposeChildSpec[];
  rationale: string;
  reason: string;
  new_title: string | null;
  new_body: string | null;
  new_assignee: string | null;
  child_ids: string[];
  prompt_tokens: number | null;
  completion_tokens: number | null;
  persisted: boolean;
}

export async function decomposeTask(taskId: string): Promise<DecomposeOutcome> {
  return apiRequest(`/kanban/tasks/${taskId}/decompose`, { method: 'POST' });
}

export async function applyDecompose(
  taskId: string,
  data: {
    fanout?: boolean;
    children?: DecomposeChildSpec[];
    new_title?: string | null;
    new_body?: string | null;
    new_assignee?: string | null;
    rationale?: string;
    prompt_tokens?: number | null;
    completion_tokens?: number | null;
  },
): Promise<DecomposeOutcome> {
  return apiRequest(`/kanban/tasks/${taskId}/apply-decompose`, {
    method: 'POST',
    body: JSON.stringify(data),
  });
}

// ==================== Pipeline Templates API ====================

export interface PipelineQuestion {
  id: string;
  type: string;
  label: string;
  options: string[];
}

export interface PipelineQuestionGroup {
  group: string;
  group_label: string;
  questions: PipelineQuestion[];
}

export interface PipelineRole {
  role_id: string;
  description: string;
  required_skills: string[];
}

export interface PipelineTaskSeed {
  title_template: string;
  description_template: string;
  role: string;
  parents: number[];
}

export interface PipelineTemplate {
  skill_id: string;
  name: string;
  description: string;
  category: string;
  tags: string[];
  task_count: number;
  roles: string[];
}

export interface PipelineTaskGraphVariant {
  id: string;
  label: string;
  description: string;
  seeds: PipelineTaskSeed[];
}

export interface PipelineTemplateDetail {
  skill_id: string;
  name: string;
  description: string;
  category: string;
  tags: string[];
  discovery_questions: PipelineQuestionGroup[];
  role_templates: PipelineRole[];
  task_graph_seed: PipelineTaskSeed[];
  task_graph_variants?: PipelineTaskGraphVariant[];
}

export interface PipelineInstantiateResult {
  task_ids: string[];
  edges: string[][];
  role_agent_mapping: Record<string, string | null>;
}

export async function listPipelines(): Promise<{ items: PipelineTemplate[]; total: number }> {
  return apiRequest('/kanban/pipelines');
}

export async function getPipelineDetail(skillId: string): Promise<PipelineTemplateDetail> {
  return apiRequest(`/kanban/pipelines/${skillId}`);
}

export async function instantiatePipeline(
  boardId: string,
  data: { skill_id: string; answers: Record<string, string>; variant_id?: string },
): Promise<PipelineInstantiateResult> {
  return apiRequest(`/kanban/boards/${boardId}/pipeline/instantiate`, {
    method: 'POST',
    body: JSON.stringify(data),
  });
}
